### ISSUE TITLE
Custom Request/Response Objects

### ISSUE THREAD

**Issue by jeffturcotte** [zendframework/zend-stratigility#23]
If I have application specific Request and Response objects that are extensions of the Diactoros PSR-7 implementations, all customization added in my extensions become unusable within my application due to the Stratigility HTTP decorators.

A solution would be to have my own custom HTTP objects be decorators themselves that decorate the Stratigility HTTP objects, but that architecture feels less than ideal. Then because of the type hints within the default FinalHandler, best practice would be to undecorate before passing the objects on.

Would it be better to have the Stratigility HTTP decorators be extensions of the Diactoros PSR-7 implementations? That way any custom Request or Response objects the programmer wanted could then extend those. I haven't looked too deeply into this yet, but after quickly browsing the code, I believe such a change would be backwards compatible.

Also, as far as I can tell, none of the added functionality of the decorated HTTP objects is used until the FinalHandler. Maybe a solution is to wait until that handler is invoked to decorate? Or perhaps the FinalHandler should be responsible for the functionality that is currently provided by the decorator?

Anyway, I'm gonna play with some ideas here, but let me know your thoughts because I feel like this problem is very limiting.

Thanks for this lib!!

**Comment by weierophinney** [comment]
@jeffturcotte You comment:

> Also, as far as I can tell, none of the added functionality of the decorated HTTP objects is used until the FinalHandler.

What you're observing there is essentially the call to `end()`. This is a nice utility, but could be omitted entirely from `FinalHandler`, and it would still work. (We'd instead do `$response->getBody()->write()` and be done; `end()` does that, but also marks the response stream immutable.)

However, that's not the reason the decorators exist. They exist for more important reasons:

- Primarily, they're to allow the end-user access to the *original* request and response (see the `getOriginalRequest()` and `getOriginalResponse()` methods of these two implementations). This is useful in nested middleware, so it can have access to the original request state - which will be the only way to get the full request URI. Which leads to the next:
- Secondarily, they're used by the `Next` implementation to help determine how to fold and unwind the URI as it traverses into and out of nested middleware. This is done using the aforementioned access to the original request/response. Nested middleware is not provided a URI with the full path, but instead the truncated path, with any ancestor paths stripped.

At this point, I want to note two important things.

First, per your comment:

> Would it be better to have the Stratigility HTTP decorators be extensions of the Diactoros PSR-7 implementations?

What I have done is have them *implement the relevant PSR-7 interfaces*. In other words, they fully follow the PSR-7 interfaces, and decorate a concrete instance of the relevant PSR-7 interface; in fact, they proxy all PSR-7 interface methods to the decorated object! As such, they can be dropped in anywhere PSR-7 interfaces are required.

Seconde, the decorators *give you access to the PSR-7 message decorated*, via `getCurrentRequest()` and `getCurrentResponse()`, respectively. Whenever you use a `with*()` method, the original PSR-7 message is operated on, and its clone is passed to a new decorator instance as well, which means you have access to everything just as you would with an undecorated message.

So, as an example, if you have created a custom concrete method `isGone()` in your response, you could do the following:

```php
if ($response->getCurrentResponse()->isGone()) {
}
```

However, one big note: that usage is a code smell, as it indicates you're depending on a concretion instead of an abstraction.

So, I'm going to flip your question back around: knowing that the decorators implement the PSR-7 interfaces, and proxy all methods to the decorated PSR-7 message, and that you have access to the decorated PSR-7 message, what exactly are you doing that would require a change?

**Comment by jeffturcotte** [comment]
Thanks @weierophinney. 

First off, I did not know I had access to the original messages. The `getCurrent*` methods are definitely a solution to my problem, so I am going to go ahead and implement with that. Usability wise, I had expected that my input would persist through the middleware as is, so that was a little shocking.

As to the rest of your post, my custom method is very similar to your `isGone` example. I guess it could be considered code smell, but by that logic, aren't `getCurrentResponse`,  `getCurrentRequest`, and all other methods provided by the decorators code smell? Using those methods within middleware creates a dependency on a concretion instead of the abstraction. :)

Philosophically, I don't see the problem with having **my own** application components depend on **my own** concretion, but with the current state of the decorators, I feel somewhat beholden to Stratigility's view of what a PSR-7 implementation looks like: A 1 to 1, by the book interpretation.

Thank you for the detailed response! Feel free to close if you have no other comments.

**Comment by weierophinney** [comment]
> There doesn't appear to be a getCurrentResponse method on the Response.

Indeed - can you open a separate PR to add that, please? That's an oversight, and it *should* be present.

(Interestingly, however, I can think of very few cases where I'd want access to the original response, unless I knew that it represented a default response to return. Most of the time, I can use the response passed to me, or create a concrete response to return from my middleware directly. The primary reason we pass the response at all is so that consuming middleware *does not need to depend on any concretions*. The main concretion of interest for responses will typically be the underlying stream representing the body: you may want to use a generator- or callback-based approach instead of the default stream-based approach.)

Now, to other questions/points you have made:

> by that logic, aren't getCurrentResponse, getCurrentRequest, and all other methods provided by the decorators code smell?

The main reason for these to exist are for the `Next` implementation. When it loops to the next in the stack, it needs to be able to get at this information in order to determine what request/response to push to the middleware. One thing to remember as you work with Stratigility (or node's Connect or ExpressJS) is that middleware may be _nested_. This allows you to compose middleware in such a way that it can be re-used.

As an example, let's create some middleware:

```php
$bar = new MiddlewarePipe();
$bar->pipe('/baz', function ($req, $res, $next) {});
$bar->pipe('/bat', function ($req, $res, $next) {});
```

Now, if we run `$bar` as the only middleware in our application, or pipe it to the top-level middleware without a path, the middleware it composes match at the URIs they specify:

```php
// only middleware in the application:
$server = Server::createServer($bar);
$server->listen();

// attached to another app at top-level:
$app = new MiddlewarePipe();
$app->pipe($bar);
$server = Server::createServer($app);
$server->listen();
```

Either of the above will now execute the middleware defined in `$bar` at `/baz` and `/bat`, respectively.

One of the key features of Stratigility is composability of middleware. Let's say we want to re-use `$bar` somewhere else. We could pipe it to its own top-level URI:

```php
$app = new MiddlewarePipe();
$app->pipe('/bar', $bar);
```

What happens now? We now match at `/bar/baz` and `/bar/bat`, respectively.

But what happens inside that middleware? What URI is present?

In order to maintain the portability of the middleware, the URI that was used during the `pipe()` invocation is stripped from the request. This is done so that matches happen at the _start_ of the URI. Additionally, it ensures that URI comparisons done internal to the middleware do not have to take into account any prefix its not aware of. As an example, let's look at a potential implementation of the `baz` middleware from above:

```php
$path = $request->getUri()->getPath();
if ($path === '/') {
    // Root of the collection!
    $response->write('In Baz collection!');
    return $response;
}

if (! preg_match('#^/(?P<id>\d+)/?$#', $path, $matches)) {
    // No identifier found!
    return $response->withStatus(404);
}

// now do something with the identifier!
$id = (int) $matches['id'];
```

If we did not truncate the URIs, this code would be unreliable:

- Finding an identifier at the end would be trivial, but
- How do we find the start of the URI that's relevant to us? (Hint: PCRE negative lookbehind is unintuitive and difficult!)

So, Stratigility keeps track of this for you via the decorators; it unwinds the last path matched to determine what depth to match at for the next in the stack, and continues. Whenever it matches a path, it strips that off and passes a _new_ request instances with the _new_ URI to the middleware invoked.

That also answers another point you raised:

> I had expected that my input messages would persist through the middleware as is.

Chances are that it will be a new instance anyways if you are doing any nesting!

Going back to the "code smell", the point I'm trying to make is that the _middleware_ attached to the system is still only hinting on PSR-7, and Stratigility is passing PSR-7 messages to the middleware. If you are truly programming to the interface, you'll only use the methods exposed in PSR-7 in your middleware. If you need convenience features, such as "isGone", those can also be done as functions, static methods, closures, etc. *without needing to rely on a concrete implementation of PSR-7*.

```php
$isGone = function (ResponseInterface $response) {
    return ($response->getStatusCode() === 410);
};

if ($isGone($response)) {
    // GONE!
}
```

(This is when I make the obligatory remark that I wish we had function autoloading available in PHP!)

> At least for the Request, the original request could've been attached to that as an attribute which would be totally in line with the PSR-7 interface,

I actually originally implemented it like that. However, that then becomes a convention-based approach, which comes with a bunch of baggage:

- Who defines the attribute key to use?
- How do I know that key exists?
- How do I know *what the value is*?
- What happens if I mix and match several middleware dispatchers — could we not have conflicts?

A type-based approach is more predictable and discoverable (it's in the API), and it's far simpler to validate:

```php
if ($request instanceof \Zend\Stratigility\Http\Request) { }

// vs

if ($request->hasAttribute('OriginalRequest')
    && $request->getAttribute('OriginalRequest') instanceof \Psr\Http\Message\RequestInterface
) { }
```

I'd hate writing that latter code more than once!

> or even better, passed directly into the FinalHandler and Next constructors within MiddlewarePipe's `__invoke` method

I tried very hard to follow a functional-programming approach with the `Next` implementation, particularly once I started refactoring it to accommodate message immutability when we made that decision for PSR-7. In Connect, you call `next()` without any arguments, or with the error. This works because the request and response are *mutable*. Once we have *immutability*, an issue arises: how do notify the dispatcher of changes so that later middleware are provided the new instances? The only way to do that reliably was to require that the invoker pass the request and response via `$next()`:

```php
return $next($request, $response);

// or, with an error:
return $next($request, $response, $error);
```

To keep `Next` stateless, while still allowing nested middleware, this meant that the request and response instances themselves needed to contain the original state somehow, and that was how the decorators came to be. (I tried a stateful approach first, and the stateless approach made the code far simpler and more consistent!)

To sum up: the decorators exist *primarily* as an implementation detail of `Next`. It's not to enforce a strict 1:1 interpretation of middleware; it's to allow *composable*, *reusable* middleware. This decision has side-effects.

The first *side-effect* is that consumers can query them for the original request/response and/or the decorated request/response. If you are programming to PSR-7, you likely will never touch these. There is, however, one very interesting use case for having them, which is for getting at the actual top-level request URI, which you may not get from your request instance normally if you are in a nested middleware instance.

The second *side-effect* is that consumers are expected to depend on the PSR-7 abstractions *not concrete implementations*. This is due to the need to decorate the instances in order to ensure the `Next` implementation works in expected ways.

I hope the above explanation of the architecture helps explain my point of view better, and the technical decisions behind it.
